contracts/authorization/src/contract.rs

```diff
@@ -803,11 +821,40 @@ fn process_polytone_callback(
                         Err(error) => {
                             if callback_info.execution_result == ExecutionResult::InProcess {
                                 callback_info.execution_result = if error == "timeout" {
-                                    ExecutionResult::Timeout
+                                    let is_expired = callback_info
+                                        .ttl
+                                        .map(|ttl| ttl.is_expired(&env.block))
+                                        .unwrap_or(true);
+                                    // If it's expired it's not retriable
+                                    ExecutionResult::Timeout(!is_expired)
                                 } else {
                                     ExecutionResult::UnexpectedError(error)
                                 };
 
+                                // If the callback is not retriable anymore, was initiated by a user and a token was sent, we'll send it back
+                                if let (
+                                    ExecutionResult::Timeout(false),
+                                    OperationInitiator::User(initiator_addr),
+                                    AuthorizationMode::Permissioned(PermissionType::WithCallLimit(
+                                        _,
+                                    )),
+                                ) = (
+                                    &callback_info.execution_result,
+                                    &callback_info.initiator,
+                                    &AUTHORIZATIONS
+                                        .load(deps.storage, callback_info.label.clone())?
+                                        .mode,
+                                ) {
+                                    let denom = build_tokenfactory_denom(
+                                        env.contract.address.as_str(),
+                                        &callback_info.label,
+                                    );
+                                    messages.push(CosmosMsg::Bank(BankMsg::Send {
+                                        to_address: initiator_addr.to_string(),
+                                        amount: coins(1, denom),
+                                    }));
+                                }
+
```

Comment:
This is written in a somewhat confusing way, the main reason is that you check if `callback_info.execution_result` is in process, change that to something else, and then check again if its ExecutionResult::Timeout, all in the same block of code.
Would be much more clear for future readers if you move the 2nd `if let` code into the first one: "if error is timeout and is expired, we refund the token and set result to be timeout expired", a much more natural flow.

Reply:
Agree, changed it to what you specified to make it more natural.